Add date opened to every update block

proteusvacuum opened the pull request:
http://manage.dimagi.com/default.asp?230043#1178916

Waiting to validate that this is the right thing to do, but opening for tests.

@benrudolph @sravfeyn

proteusvacuum committed:
Add date opened to every update and create block

dimagimon commented on corehq/ex-submodules/casexml/apps/case/mock.py:
```diff
@@ -115,10 +117,9 @@ def __init__(self,
-        })
+
         create_or_update = {
+            'date_opened':                  date_opened,
```
E241 multiple spaces after ':'

proteusvacuum replied:
yeah, no kidding.

snopoke replied:
lol

proteusvacuum commented:
OK, got feedback. I was almost there, will reopen tomorrow.

proteusvacuum committed:
Change from docstring to comment

proteusvacuum committed:
Include date_opened in every update block

dimagimon commented on corehq/ex-submodules/casexml/apps/case/tests/test_rebuild.py:
```diff
@@ -106,11 +107,13 @@ def test_couch_soft_rebuild(self):
         self.assertEqual(case.p3, 'p3-2') # new
-        self.assertEqual(3, len(case.actions)) # create + 2 updates
-        a1 = case.actions[1]
+        self.assertEqual(4, len(case.actions)) # create + update (2 actions) + 2 updates
```
E261 at least two spaces before inline comment

dimagimon commented on corehq/ex-submodules/casexml/apps/case/mock.py:
```diff
@@ -137,6 +137,7 @@ def __init__(self,
             })
         self['update'].update({
             'external_id':              external_id,
+            'date_opened':              date_opened,
```
E241 multiple spaces after ':'

proteusvacuum committed:
Fix some tests for adding date_opened to all update blocks

proteusvacuum committed:
remove datetime formatting in api test

proteusvacuum committed:
Change order of date opened in mock

proteusvacuum committed:
change order of date opened in claim endpoint test

proteusvacuum committed:
Add date_opened kwarg to update test

proteusvacuum commented:
:open_mouth: :arrow_right:  :white_check_mark:

proteusvacuum commented on corehq/apps/cloudcare/tests/test_api.py:
```diff
@@ -304,13 +303,13 @@ def _create_case(user, type, close=False, **extras):
         user_id=user.user_id,
         owner_id=user.user_id,
         **extras
-    ).as_xml(format_datetime=date_to_xml_string)]
```
I'm not really sure why this was here, and removing it might mean that something else is broken?

benrudolph replied:
i think it actually has the same functionality. date_to_xml_string just calls json_format_date which is the default `format_date` function for `as_xml` with the one difference being that `date_to_xml_string` allows for `None` dates

proteusvacuum commented:
Lies! Travis didn't run.

benrudolph commented:
haha that would've been amazing

czue commented:
this is fun.

proteusvacuum commented:
```py
False
```

proteusvacuum committed:
Comment not docstring

proteusvacuum committed:
Search for form id in all actions

CaseTransactions and CaseActions have different semantics, so using
specific array indices doesn't make sense with the
@run_with_all_backends decorator

proteusvacuum committed:
Check that the ledger transaction action happens

proteusvacuum committed:
Allow case to have date_opened in every update block

proteusvacuum commented:
huzzah! :white_check_mark: 
new code buddy: @millerdev

millerdev commented on corehq/apps/commtrack/tests/test_rebuild.py:
```diff
@@ -90,16 +90,14 @@ def test_inferred(self):
         case = CaseAccessors(self.domain).get_case(self.case.case_id)
         self.assertEqual(case.xform_ids[1:], [form_id])
-        self.assertEqual(case.actions[1].form_id, form_id)
+        self.assertTrue(form_id in [action.form_id for action in case.actions])
```
no need to change, but could use `self.assertIn(form_id, [action.form_id for action in case.actions])`

czue commented:
nice. actually not as bad as i thought this might be.

snopoke commented on corehq/ex-submodules/casexml/apps/case/mock.py:
```diff
@@ -99,7 +99,9 @@ def __init__(self,
-        date_modified = date_modified or datetime.utcnow()
+        now = datetime.utcnow()
+        date_modified = date_modified or now
+        date_opened = date_opened or date_modified
```
bumping my head against this now. I'm confused why you'd want to include 'date_opened' in the update block for a case that is already open?

I understand the need for it in the restore but this class is used for generating case blocks that go into forms.

czue replied:
@snopoke I think that's an oversight. we should probably only specify `date_opened` if the case is also being created.

snopoke replied:
https://github.com/dimagi/commcare-hq/pull/13112
